### DISCUSSION TITLE
Is this Terraform provider a relevant way to manage Power Platform solutions?

### DISCUSSION THREAD

**Discussion by rpothin** [microsoft/terraform-provider-power-platform#297] [General]
Hi team 👋🏼

First, thank you for your great work with this Terraform provider for Power Platform 👏🏼
It is a really interesting approach to manage Power Platform key resources to have a healthy platform (_billing policies, DLP policies, environments, tenant settings..._).

In the other hand, I am not sure solutions management should be in the scope of this Terraform provider 🤔
There are at least 3 different ways to manage this:
- [Pipelines in Power Platform](https://learn.microsoft.com/en-us/power-platform/alm/pipelines) for makers (_low code / no code developers_)
- [Microsoft Power Platform Build Tools for Azure DevOps](https://learn.microsoft.com/en-us/power-platform/alm/devops-build-tools) or [GitHub Actions for Microsoft Power Platform](https://learn.microsoft.com/en-us/power-platform/alm/devops-github-actions) for code-first developers
- [Power Platform CLI](https://learn.microsoft.com/en-us/power-platform/developer/cli/reference/) for code-first developers who used a CI/CD platform not covered above or for more exotic scenarios

From my perspective, infrastructure as code is more for the management of the underlying services supporting applications, but the applications configuration / code should be managed in a different way.

Also, from an adoption perspective, I think it will be easier to get some interest from administration / governance teams to use infrastructure as code to manage the platform itself than to try to get development teams also use it for the deployment of solutions.

What do you think?

Have a great week.
Regards,
Raphael

**Comment by mawasile** [comment]
Hi @rpothin 

We had that discussion some time ago internally, and had/have the same feal about it. The reason that we left solutions in the provider is that we see patterns by customer where setting up the whole environment/tenant consist also from importing solution and setting custom PP/Dynamics based customer product including configuring some env settings, env variables, connections etc. 

That is also the reason why we (for example) implementing [powerplatform_data_record](https://github.com/microsoft/terraform-provider-power-platform/issues/278) but this one will be limited on purpose without any performance optimizations and with limited amount of records it can work with. 

This provider is created to help with administrative tasks, infrastructure observability, monitoring and security and not to replace end to end ALM process. Could be, that we need to be more explicit about it in the documentation, but that is also part of best practices of one applying ALM in the tenant. 

Cheers,
Mateusz

**Reply by rpothin** [reply]
Hi @mawasile,

Thank you for your answer.

I will be curious to have some real-world examples where it would make more sense to use the Terraform provider for an initial import of solutions in place of another more usual CI/CD approach (_like the ones mentioned in my initial comment_) 🤔
Are you talking here about only an initial import of solutions or also managing updates of these solutions still using the Terraform provider?

I am 100% aligned with the values (_administrative tasks, infrastructure observability, monitoring and security_) you mentioned in your comment where this Terraform provider could definitely help the Power Platform community.

Regarding the idea of being more explicit in the documentation regarding ALM, I definitely support it.
Perhaps some sort of "disclaimer", mentioning that the `powerplatform_solution (Resource)` can be used in some scenarios (_giving a few examples_) and then pointing to the main ALM approaches recommended and supported by Microsoft, could be a good way to clarify this part.

Have a great day.
Raphael

**Reply by mattdot** [reply]
Great discussion.  This is not unique to just the Power Platform Terraform Provider.  In many cases the line between what is considered Infrastructure and what is considered an application is kind of fuzzy.  The Power Platform Terraform Provider provides the ability to deploy solutions, but it is not intended to replace ALM processes **for the applications that your team maintains**.  We still think that Power Pipelines and CoE Toolkit are great options for implementing your ALM processes (and we're working with those teams to ensure seamless integration).  However, as the owner of the application / infrastructure for your scenario, you are in the best position to decide what is part of your application and what is part of your infrastructure.  I personally default to using ALM for things where my team maintains the code (or low-code) assets.

There may be solutions installed from AppSource or as 3rd-party managed solutions that you install into your environment that you treat more like the foundation that your app is built on top of.  There are quite a few security and compliance solutions that could fall into this area.  Similarly, you might have some enterprise-wide solutions that you install into each environment.  Even the core dynamics products like Customer Service are kind of Microsoft apps that you may treat as "infrastructure".

**Reply by rpothin** [reply]
Hi @mattdot,

Thank you very much for sharing your vision about this.
I find it really interesting and I can see where you are going with this approach.

As a member of the platform team in charge of Power Platform, I definitely see the value of being able to setup environments based on templates with pre-defined settings, application users, solutions...

To make it real and to allow people to understand this vision I think it should be (_even if I am pretty sure it is already in your plan_) one of the modules provided in the Terraform registry mentioned in @mawasile [comment in another discussion](https://github.com/microsoft/terraform-provider-power-platform/discussions/365#discussioncomment-10064140).

Thank you both for your clarification on this topic, I think we are aligned.
I am looking forward to seeing the next chapters of this adventure.

**Comment by webstean** [comment]
I agree that this provider isn't the answer to ALM, there are other capabilities better suited to that. I see this provider's use-case, to be basically around the setup and maintenance of the entire Power App environments, including environment groups, environments, access control and tenant settings, especially the DLP settings, VNet integration, Identity and CMK encryption.
